File: backend/app/api/contracts.py
"""
API endpoints for contract interactions
"""
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from typing import Optional
from app.services.contract_service import (
    vault_service,
    session_key_service,
    position_service,
    rebalance_service
)

router = APIRouter(prefix="/api/contracts", tags=["contracts"])


class VaultBalanceRequest(BaseModel):
    user_address: str


class VaultDepositRequest(BaseModel):
    user_address: str
    amount: int


class SessionKeyCreateRequest(BaseModel):
    user_address: str
    expiry_days: int = 30


class SessionKeyValidateRequest(BaseModel):
    session_key: str


class PositionOpenRequest(BaseModel):
    user_address: str
    pool_id: str
    amount: int
    min_price: int
    max_price: int


class PositionGetRequest(BaseModel):
    position_id: str


@router.post("/vault/balance")
async def get_vault_balance(request: VaultBalanceRequest):
    """Get user's vault balance from contract"""
    try:
        balance = await vault_service.get_balance(request.user_address)
        
        if balance is None:
            return {"success": False, "error": "Failed to fetch balance"}
        
        return {
            "success": True,
            "balance": balance,
            "balance_formatted": f"{balance / 1e18:.6f} ETH"
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/vault/deposit")
async def deposit_to_vault(request: VaultDepositRequest):
    """
    Initiate deposit transaction (frontend will sign with wallet)
    Returns transaction data for signing
    """
    try:
        return {
            "success": True,
            "transaction": {
                "contract_address": vault_service.contract_address,
                "entrypoint": "deposit",
                "calldata": [hex(request.amount)]
            },
            "message": "Transaction prepared. Please sign with your wallet."
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/vault/withdraw")
async def withdraw_from_vault(request: VaultDepositRequest):
    """
    Initiate withdrawal transaction
    Returns transaction data for signing
    """
    try:
        return {
            "success": True,
            "transaction": {
                "contract_address": vault_service.contract_address,
                "entrypoint": "withdraw",
                "calldata": [hex(request.amount)]
            },
            "message": "Transaction prepared. Please sign with your wallet."
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/session-key/validate")
async def validate_session_key(request: SessionKeyValidateRequest):
    """Check if a session key is valid"""
    try:
        is_valid = await session_key_service.is_valid(request.session_key)
        
        if is_valid:
            permissions = await session_key_service.get_permissions(request.session_key)
            return {
                "success": True,
                "valid": True,
                "permissions": permissions
            }
        else:
            return {
                "success": True,
                "valid": False
            }
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/session-key/create")
async def create_session_key(request: SessionKeyCreateRequest):
    """
    Initiate session key creation transaction
    Returns transaction data for signing
    """
    try:
        return {
            "success": True,
            "transaction": {
                "contract_address": session_key_service.contract_address,
                "entrypoint": "create_session_key",
                "calldata": [hex(request.expiry_days)]
            },
            "message": "Session key creation prepared. Please sign with your wallet."
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/position/get")
async def get_position(request: PositionGetRequest):
    """Get position details from contract"""
    try:
        position = await position_service.get_position(request.position_id)
        
        if position is None:
            return {"success": False, "error": "Position not found"}
        
        return {
            "success": True,
            "position": position
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/position/open")
async def open_position(request: PositionOpenRequest):
    """
    Initiate position opening transaction
    Returns transaction data for signing
    """
    try:
        return {
            "success": True,
            "transaction": {
                "contract_address": position_service.contract_address,
                "entrypoint": "open_position",
                "calldata": [
                    request.pool_id,
                    hex(request.amount),
                    hex(request.min_price),
                    hex(request.max_price)
                ]
            },
            "message": "Position opening prepared. Please sign with your wallet."
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/info")
async def get_contract_info():
    """Get deployed contract addresses"""
    from app.config import settings
    
    return {
        "success": True,
        "contracts": {
            "vault": settings.STARKNET_VAULT_CONTRACT,
            "session_key": settings.STARKNET_SESSION_KEY_CONTRACT,
            "position": settings.STARKNET_POSITION_CONTRACT,
            "rebalance": settings.STARKNET_REBALANCE_CONTRACT
        },
        "network": {
            "rpc_url": settings.STARKNET_RPC_URL,
            "chain_id": settings.STARKNET_CHAIN_ID,
            "network": settings.STARKNET_NETWORK
        }
    }

File: frontend/src/lib/contracts.js
/**
 * Contract interaction utilities
 * Handles communication between frontend, backend, and Starknet contracts
 */
import api from './api'

/**
 * Get deployed contract addresses and network info
 */
export async function getContractInfo() {
  try {
    const response = await api.get('/api/contracts/info')
    return response.data
  } catch (error) {
    console.error('Failed to get contract info:', error)
    throw error
  }
}

/**
 * Get vault balance for a user address
 */
export async function getVaultBalance(userAddress) {
  try {
    const response = await api.post('/api/contracts/vault/balance', {
      user_address: userAddress
    })
    return response.data
  } catch (error) {
    console.error('Failed to get vault balance:', error)
    throw error
  }
}

/**
 * Deposit funds to vault
 * @param {string} userAddress - User's Starknet address
 * @param {string} amount - Amount in wei (as string)
 * @param {Function} executeTransaction - Function from WalletContext to execute transaction
 */
export async function depositToVault(userAddress, amount, executeTransaction) {
  try {
    // Get transaction data from backend
    const response = await api.post('/api/contracts/vault/deposit', {
      user_address: userAddress,
      amount: parseInt(amount)
    })

    if (!response.data.success) {
      throw new Error(response.data.error || 'Failed to prepare deposit')
    }

    const { transaction } = response.data

    // Execute transaction through wallet
    const result = await executeTransaction(
      transaction.contract_address,
      transaction.entrypoint,
      transaction.calldata
    )

    return result
  } catch (error) {
    console.error('Deposit failed:', error)
    throw error
  }
}

/**
 * Withdraw funds from vault
 */
export async function withdrawFromVault(userAddress, amount, executeTransaction) {
  try {
    const response = await api.post('/api/contracts/vault/withdraw', {
      user_address: userAddress,
      amount: parseInt(amount)
    })

    if (!response.data.success) {
      throw new Error(response.data.error || 'Failed to prepare withdrawal')
    }

    const { transaction } = response.data

    const result = await executeTransaction(
      transaction.contract_address,
      transaction.entrypoint,
      transaction.calldata
    )

    return result
  } catch (error) {
    console.error('Withdrawal failed:', error)
    throw error
  }
}

/**
 * Create session key
 */
export async function createSessionKey(userAddress, expiryDays, executeTransaction) {
  try {
    const response = await api.post('/api/contracts/session-key/create', {
      user_address: userAddress,
      expiry_days: expiryDays || 30
    })

    if (!response.data.success) {
      throw new Error(response.data.error || 'Failed to prepare session key creation')
    }

    const { transaction } = response.data

    const result = await executeTransaction(
      transaction.contract_address,
      transaction.entrypoint,
      transaction.calldata
    )

    return result
  } catch (error) {
    console.error('Session key creation failed:', error)
    throw error
  }
}

/**
 * Validate session key
 */
export async function validateSessionKey(sessionKey) {
  try {
    const response = await api.post('/api/contracts/session-key/validate', {
      session_key: sessionKey
    })
    return response.data
  } catch (error) {
    console.error('Session key validation failed:', error)
    throw error
  }
}

/**
 * Open trading position
 */
export async function openPosition(params, executeTransaction) {
  try {
    const response = await api.post('/api/contracts/position/open', {
      user_address: params.userAddress,
      pool_id: params.poolId,
      amount: params.amount,
      min_price: params.minPrice,
      max_price: params.maxPrice
    })

    if (!response.data.success) {
      throw new Error(response.data.error || 'Failed to prepare position opening')
    }

    const { transaction } = response.data

    const result = await executeTransaction(
      transaction.contract_address,
      transaction.entrypoint,
      transaction.calldata
    )

    return result
  } catch (error) {
    console.error('Position opening failed:', error)
    throw error
  }
}

/**
 * Get position details
 */
export async function getPosition(positionId) {
  try {
    const response = await api.post('/api/contracts/position/get', {
      position_id: positionId
    })
    return response.data
  } catch (error) {
    console.error('Failed to get position:', error)
    throw error
  }
}

/**
 * Convert ETH amount to wei
 */
export function ethToWei(ethAmount) {
  return (BigInt(Math.floor(ethAmount * 1e18))).toString()
}

/**
 * Convert wei to ETH
 */
export function weiToEth(weiAmount) {
  return Number(BigInt(weiAmount)) / 1e18
}
